include/arch/arm/aarch32/arch.h

```diff
@@ -150,98 +150,56 @@ extern "C" {
- *
- * This is used to determine the required alignment of threads'
- * privilege stacks when building with support for user mode.
- *
- * @note
- * The privilege stacks do not need to respect the minimum MPU
- * region alignment requirement (unless this is enforced via
- * the MPU Stack Guard feature).
- */
-#if defined(CONFIG_USERSPACE)
-#define Z_PRIVILEGE_STACK_ALIGN MAX(ARCH_STACK_PTR_ALIGN, Z_MPU_GUARD_ALIGN)
-#endif
-
-/**
- * @brief Calculate power of two ceiling for a buffer size input
- *
- */
-#define POW2_CEIL(x) ((1 << (31 - __builtin_clz(x))) < x ?  \
-		1 << (31 - __builtin_clz(x) + 1) : \
-		1 << (31 - __builtin_clz(x)))
-
 #if defined(CONFIG_USERSPACE) && \
 	defined(CONFIG_MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT)
-/* Guard is 'carved-out' of the thread stack region, and the supervisor
- * mode stack is allocated elsewhere by gen_priv_stack.py
+/* This MPU requires regions to be sized to a power of two, and aligned to
+ * their own size. Since an MPU region must be able to cover the entire
+ * user-accessible stack buffer, we size/align to match. The privilege
+ * mode stack is generated elsewhere in memory.
  */
-#define ARCH_THREAD_STACK_RESERVED 0
-#else
-#define ARCH_THREAD_STACK_RESERVED MPU_GUARD_ALIGN_AND_SIZE
-#endif
-
-#if defined(CONFIG_USERSPACE) && \
-	defined(CONFIG_MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT)
-#define ARCH_THREAD_STACK_DEFINE(sym, size) \
-	struct z_thread_stack_element __noinit \
-		__aligned(POW2_CEIL(size)) sym[POW2_CEIL(size)]
+#define ARCH_THREAD_STACK_OBJ_ALIGN(size)	Z_POW2_CEIL(size)
+#define ARCH_THREAD_STACK_SIZE_ADJUST(size)	Z_POW2_CEIL(size)
 #else
-#define ARCH_THREAD_STACK_DEFINE(sym, size) \
-	struct z_thread_stack_element __noinit __aligned(STACK_ALIGN) \
-		sym[size+MPU_GUARD_ALIGN_AND_SIZE]
+#define ARCH_THREAD_STACK_OBJ_ALIGN(size)	MAX(Z_THREAD_MIN_STACK_ALIGN, \
+						    Z_MPU_GUARD_ALIGN)
+#ifdef CONFIG_USERSPACE
+#define ARCH_THREAD_STACK_SIZE_ADJUST(size) \
+	ROUND_UP(size, CONFIG_ARM_MPU_REGION_MIN_ALIGN_AND_SIZE)
 #endif
-
-#if defined(CONFIG_USERSPACE) && \
-	defined(CONFIG_MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT)
-#define ARCH_THREAD_STACK_LEN(size) (POW2_CEIL(size))
-#else
-#define ARCH_THREAD_STACK_LEN(size) ((size)+MPU_GUARD_ALIGN_AND_SIZE)
 #endif
 
-#if defined(CONFIG_USERSPACE) && \
-	defined(CONFIG_MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT)
-#define ARCH_THREAD_STACK_ARRAY_DEFINE(sym, nmemb, size) \
-	struct z_thread_stack_element __noinit \
-		__aligned(POW2_CEIL(size)) \
-		sym[nmemb][ARCH_THREAD_STACK_LEN(size)]
-#else
-#define ARCH_THREAD_STACK_ARRAY_DEFINE(sym, nmemb, size) \
-	struct z_thread_stack_element __noinit \
-		__aligned(STACK_ALIGN) \
-		sym[nmemb][ARCH_THREAD_STACK_LEN(size)]
+#ifdef CONFIG_MPU_STACK_GUARD
+/* Kernel-only stacks need an MPU guard region programmed at the beginning of
+ * the stack object, so align the object appropriately.
+ */
+#define ARCH_KERNEL_STACK_RESERVED	MPU_GUARD_ALIGN_AND_SIZE
+#define ARCH_KERNEL_STACK_OBJ_ALIGN	Z_MPU_GUARD_ALIGN
```

Comment:
@andrewboie why is this not MAX(MAX(Z_THREAD_MIN_STACK_ALIGN, Z_MPU_GUARD_ALIGN) ??

Reply:
It's like this because Z_MPU_GUARD_ALIGN is always equal to or larger than Z_THREAD_MIN_STACK_ALIGN.

Z_THREAD_MIN_STACK_ALIGN is defined as "Denotes the minimum required alignment of a thread stack." (not necessarily a kernel-only stack) and is either defined to the smallest MPU region (if userspace) or the stack pointer alignment (if not).

Z_MPU_GUARD_ALIGN is always the smallest MPU region necessary to program a guard:
```
#if defined(CONFIG_MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT)
#define Z_MPU_GUARD_ALIGN (MAX(MPU_GUARD_ALIGN_AND_SIZE, \
	MPU_GUARD_ALIGN_AND_SIZE_FLOAT))
#else
#define Z_MPU_GUARD_ALIGN MPU_GUARD_ALIGN_AND_SIZE
#endif
```

Z_MPU_GUARD_ALIGN is, by design, supposed to capture the necessary alignment to instantiate a stack guard at the lowest address. If it doesn't there is a problem with its definition.

Reply:
That's correct thanks. Hadn't seen the #ifdef CONFIG_MPU_STACK_GUARD on top.
By the way, why there is no #else here? Don't you want to have these arch specific defines in Arm arch.c when no guard feature is enabled ?

Reply:
> By the way, why there is no #else here? Don't you want to have these arch specific defines in Arm arch.c when no guard feature is enabled ?

I'm not sure I understand what you mean.
If `CONFIG_MPU_STACK_GUARD` is not enabled, we don't need the other definitions like `Z_MPU_GUARD_ALIGN` as there are no stack guards programmed. All the C code logic for guards is also #ifdef'ed around `CONFIG_MPU_STACK_GUARD`. I could add #else cases defined as 0 I suppose but nothing would be using it.

The core kernel only uses the definitions prefixed with `ARCH_`.

Reply:
It's fair to say that a lot of the intermediate definitions in this header are just feeding into getting the `ARCH_` definitions set right. This patch simplifies arch.h somewhat but there's probably more cleanup and proper namespacing that could be done.

Reply:
Well, I meant to have something like: 
```
#else /* !CONFIG_MPU_STACK_GUARD */
#define ARCH_KERNEL_STACK_OBJ_ALIGN	ARCH_STACK_PTR_ALIGN
#endif
```

I guess this is not an issue because if ARCH_KERNEL_STACK_OBJ_ALIGN is not defined then Z_KERNEL_STACK_OBJ_ALIGN is already set to ARCH_STACK_PTR_ALIGN the kernel header. But for ease of reading it would be nice if the arm arch.h already defines the ARCH_KERNEL_STACK_OBJ_ALIGN in all scenarios.